zebra-rpc/src/methods/get_block_template_rpcs/types/long_poll.rs

```diff
@@ -0,0 +1,238 @@
+    pub tip_height: Height,
+
+    /// If the tip block hash changes, a new template must be provided.
+    /// Old work is no longer valid.
+    pub tip_hash: block::Hash,
+
+    /// If the max time is reached, a new template must be provided.
+    /// Old work is no longer valid.
+    ///
+    /// Ideally, a new template should be provided at least one target block interval before
+    /// the max time. This avoids wasted work.
+    pub max_time: DateTime<Utc>,
+
+    // Fields that allow old work:
+    //
+    /// If the mempool transactions change, a new template might be provided.
+    /// Old work is still valid.
+    ///
+    /// We ignore changes to authorizing data.
+    pub mempool_transaction_mined_ids: Arc<[transaction::Hash]>,
+}
+
+impl LongPollInput {
+    /// Returns a new [`LongPollInput`], based on the supplied fields.
+    pub fn new(
+        tip_height: Height,
+        tip_hash: block::Hash,
+        max_time: DateTime<Utc>,
+        mempool_tx_ids: impl IntoIterator<Item = UnminedTxId>,
+    ) -> Self {
+        let mempool_transaction_mined_ids =
+            mempool_tx_ids.into_iter().map(|id| id.mined_id()).collect();
+
+        LongPollInput {
+            tip_height,
+            tip_hash,
+            max_time,
+            mempool_transaction_mined_ids,
+        }
+    }
+}
+
+/// The encoded long poll ID, generated from the [`LongPollInput`].
+///
+/// `zcashd` IDs are currently 69 hex/decimal digits long.
+/// Since Zebra's IDs are only 46 hex/decimal digits, mining pools should be able to handle them.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(try_from = "String", into = "String")]
+pub struct LongPollId {
+    // Fields that invalidate old work:
+    //
+    /// If the tip block height changes, a new template must be provided.
+    /// Old work is no longer valid.
+    ///
+    /// We use the height as well as the hash, to reduce the probability of a missed tip change.
+    pub tip_height: u32,
+
+    /// If the tip block changes, a new template must be provided.
+    /// Old work is no longer valid.
+    /// This checksum is not cryptographically secure.
+    ///
+    /// It's ok to do a probabilistic check here,
+    /// so we choose a 1 in 2^32 chance of missing a block change.
+    pub tip_hash_checksum: u32,
+
+    /// If the max time is reached, a new template must be provided.
+    /// Old work is no longer valid.
+    ///
+    /// Ideally, a new template should be provided at least one target block interval before
+    /// the max time. This avoids wasted work.
+    ///
+    /// Zcash times are limited to 32 bits by the consensus rules.
+    pub max_timestamp: u32,
+
+    // Fields that allow old work:
+    //
+    /// If the number of mempool transactions changes, a new template might be provided.
+    /// Old work is still valid.
+    ///
+    /// The number of transactions is limited by the mempool DoS limit.
+    pub mempool_transaction_count: u32,
```

Comment:
This could be a `u16` if we want, because the number of transactions in the main pool is usually small. And it doesn't matter if we wrap when converting from `usize`, because we're only trying to detect changes. (The gap between identical wrapped `u16`s is 2<sup>16</sup>.)

But consistently using `u32`s is fine too.